### ISSUE TITLE
CVE-2021-42574

### ISSUE THREAD

**Issue by jonathanstowe** [rakudo/rakudo#4612]
I think it's prudent to consider https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-42574

The CVE came out of embargo today so I've had a quick look at it.

TLDR;  An exploit can use the unicode bi-directional text control characters to insert code that is compiled differently to how it may appear in some tools.

The offending characters appear to be:

U+202A : LEFT-TO-RIGHT EMBEDDING
U+202B : RIGHT-TO-LEFT EMBEDDING
U+202C : POP DIRECTIONAL FORMATTING
U+202D : LEFT-TO-RIGHT OVERRIDE
U+202E : RIGHT-TO-LEFT OVERRIDE
U+2066 : LEFT-TO-RIGHT ISOLATE
U+2067 : RIGHT-TO-LEFT ISOLATE
U+2068 : FIRST STRONG ISOLATE
U+2069 : POP DIRECTIONAL ISOLATE

That may not be exhaustive, just what I gleaned from various places.

So far I have determined that the characters can't be inserted into arbitrary places in raku code (because they are in the punctuation group not whitespace,) and can't used in an identifier.

They can however be used to define both operators and terms which means they can then be placed anywhere that an operator or term can be used. They can also be used in a quote construct.

I've yet to come up with an example that will display as plausible and correct-looking code which actually compiles to do something else but I think that's for want of trying hard enough rather than it not being possible.

Obviously the most extreme solution would be to forbid the use of those characters in raku source, possibly providing some escape for use in quoted strings (though the \x form is available.)

**Comment by jonathanstowe** [comment]
![Screenshot from 2021-11-01 16-37-55](https://user-images.githubusercontent.com/736419/139708717-5a401245-7f5e-4b83-a09b-405a6a350bd6.png)

Struggling to find a way of showing the *actual* source here.

**Comment by nxadm** [comment]
Just an addition to the discussion: https://research.swtch.com/trojan

**Comment by Altai-man** [comment]
>Obviously the most extreme solution would be to forbid the use of those characters in raku source, possibly providing some escape for use in quoted strings (though the \x form is available.)

By all means this.

**Comment by jnthn** [comment]
Would outright forbidding them be hostile to folks legitimately writing string literals in RTL languages, though?

**Comment by jonathanstowe** [comment]
As I understand it an editor that is inserting  RTL text (with an RTL keyboard layout,) just inserts the characters from right to left rather than inserting the control characters and putting them LTR, but that could be just the editor/input device I tested with.  So if this is the case it would only affect trying to put a script that is usually LTR as RTL.

**Comment by moon-chilled** [comment]
How about: ltr/rtl switch is allowed, but only in comments and strings, and it must be terminated before the end of the comment/string.  That would ensure that no text appears to be quoted/commented but is not.

**Comment by jonathanstowe** [comment]
It is also possible to introduce these as operators and terms, I've yet to make one that renders as plausible code but that's not to say it is impossible, just that I haven't tried hard enough.

**Comment by moon-chilled** [comment]
I should be clearer: I propose literal directional switches not be allowed in any context other than comments and strings (with the limitations I mentioned).  So you will not be allowed create an operator or term with directional switches.

Heavier-weight grammer modifications may make them available.  However, since regex cannot express them literally, they will have to use \x or similar, giving away the bacon.

**Comment by jonathanstowe** [comment]
> not be allowed create an operator or term with directional switches.

Yeah, it's not a particularly useful thing to be able to do as, depending on one's editor, they will be invisible regardless of the other potential side-effects.

**Comment by lizmat** [comment]
How about just allowing these codepoints with a lexical pragma?
